modules/tinymce/src/core/test/ts/browser/paste/PasteTest.ts

```diff
@@ -491,13 +491,40 @@ describe('browser.tinymce.core.paste.PasteTest', () => {
       const editor = hook.editor();
       const beforeinputEvent = Singleton.value<EditorEvent<InputEvent>>();
       const inputEvent = Singleton.value<EditorEvent<InputEvent>>();
+      const setBeforeInputEvent = (e: EditorEvent<InputEvent>) => beforeinputEvent.set(e);
+      const setInputEvent = (e: EditorEvent<InputEvent>) => inputEvent.set(e);
 
-      editor.on('beforeinput', (e) => beforeinputEvent.set(e));
-      editor.on('input', (e) => inputEvent.set(e));
+      editor.on('beforeinput', setBeforeInputEvent);
+      editor.on('input', setInputEvent);
 
       const html = '<p>Test</p>';
       editor.execCommand('mceInsertClipboardContent', false, { html });
       await PasteEventUtils.pWaitForAndAssertInputEvents(beforeinputEvent, inputEvent, html);
+      TinyAssertions.assertContent(editor, '<p>Test</p>');
+
+      editor.off('beforeinput', setBeforeInputEvent);
+      editor.off('input', setInputEvent);
+    });
+
+    it('TINY-9829: Paste can be cancelled by beforeinput event', async () => {
+      const editor = hook.editor();
+      const cancelInputEvent = (e: EditorEvent<InputEvent>) => {
+        e.preventDefault();
+      };
+      const inputEvent = Singleton.value<EditorEvent<InputEvent>>();
+      const setInputEvent = (e: EditorEvent<InputEvent>) => inputEvent.set(e);
+
+      editor.on('beforeinput', cancelInputEvent);
+      editor.on('input', setInputEvent);
+
+      const html = '<p>Test</p>';
+      editor.execCommand('mceInsertClipboardContent', false, { html });
```

Comment:
I think this is just as readable if you need the `html` variable. I assume this is pasted from the code above where the `html` is used multiple times were it makes more sense.

```suggestion
      editor.execCommand('mceInsertClipboardContent', false, { html: '<p>Test</p>'' });
```